CASSANDRA-18355 Avoid serializing full TxnData instances to Accord state tables

maedhroz opened the pull request:
patch by Caleb Rackliffe; reviewed by ? for CASSANDRA-18355

maedhroz committed:
CEP-15: Multi-Partition Transaction CQL Support (Alpha)

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-17719

Co-authored-by: Caleb Rackliffe <calebrackliffe@gmail.com>
Co-authored-by: Blake Eggleston <bdeggleston@gmail.com>
Co-authored-by: David Capwell <dcapwell@apache.org>

belliottsmith committed:
Partial Replication of Transaction State

The replication of transactions and their processing state must be optimal, i.e. linearly proportional to the size of the transaction and the number of keys involved. This patch ensures a transaction is replicated in slices, with each slice being written (replication factor) times only to the shard that owns the keys, and only once (except in the case the transaction is recovered).

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-17109

bdeggleston committed:
CEP-15 (C*): Messaging and storage engine integration

patch by Blake Eggleston; reviewed by Benedict Elliott Smith, David Capwell for CASSANDRA-17103

belliottsmith committed:
CEP-15: Routables
  - Integrate accord-core changes for CASSANDRA-18087

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18087

dcapwell committed:
Command table now uses a local version added to the payload rather than rely on a version column, and fixed a few feedback related issues

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-17103

belliottsmith committed:
Transaction Expiration

Ensure old or abandoned transactions are invalidated wherever possible.

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18041

belliottsmith committed:
Invalidation fixes/improvements
 - Integrate accord-core changes for CASSANDRA-18057

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18057

belliottsmith committed:
Refactor response tracking to improve efficiency and clarity; introduce dedicated property tests; re-activate fast-path during range movements

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18056

aweisberg committed:
Support CAS and serial read on Accord

patch by Ariel Weisberg; Reviewed by Caleb Rackliffe for CASSANDRA-18100

dcapwell committed:
Refactor AccordTestBase to block retries on non-idempotent transactions.
Some tests may be flaky now due to Preempted being thrown.

belliottsmith committed:
Shard local CommandStores on contiguous ranges

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18142

belliottsmith committed:
Refactor Timestamp/TxnId
 - Combine real and logical into a single 64-but HLC
 - Introduce 16 flag bits
 - Pack epoch (48-bits), HLC (64-bits) and flags (16-bits) into two longs in memory

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18172

belliottsmith committed:
Introduce RangeDeps

Refactor Deps into KeyDeps and RangeDeps

patch by Benedict; reviewed by Aleksey Yeschenko for CASSANDRA-18173

belliottsmith committed:
Introduce Range transactions

patch by Benedict; reviewed by Ariel Weisberg for CASSANDRA-18174

belliottsmith committed:
ninja: pick a cassandra-accord SHA to build against to decouple source trees

dcapwell committed:
CASSANDRA-18154: CEP-15: Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

dcapwell committed:
CEP-15: (C*) Enhance returning SELECT to allow partition and clustering IN clauses to return multiple partitions/rows

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18154

maedhroz committed:
Post-rebase fixes around CASSANDRA-18099 and CASSANDRA-18114 (among other things)

bdeggleston committed:
add AsyncChain implementations and tests

Patch by Blake Eggleston; Reviewed by David Capwell & Benedict Elliott Smith for CASSANDRA-18004

maedhroz committed:
Feature Flag for Accord Transactions

patch by Caleb Rackliffe; reviewed by David Capwell for CASSANDRA-18195

iamaleksey committed:
Switch Node.Id from long to int

patch by Aleksey Yeschenko; reviewed by Benedict Elliott Smith for
CASSANDRA-18135

dcapwell committed:
Ninja: CASSANDRA-18214 added rat checks to Accord, but our metadata files .BRANCH and .REPO do not have a license, causing the build to fail; exclude rat when building accord

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18214

maedhroz committed:
use equals() rather than reference equality for excluding ourselves from the dependency builder

patch by Benedict; reviewed by Caleb Rackliffe for CASSANDRA-18218

bdeggleston committed:
CEP-15/Accord Immutable State Refactor

Patch by Blake Eggleston; Reviewed by David Capwell and Benedict Elliott Smith for Cassandra-18192

jacek-lewandowski committed:
Fix null value handling for static columns

patch by <jacek-lewandowski>; reviewed by <maedhroz> and <dcapwell> for CASSANDRA-18241

iamaleksey committed:
CEP-15: Store PreAccept, Accept, Commit, and Apply messages
in a durable log before processing by CommandStores

patch by Aleksey Yeschenko; reviewed by David Capwell for
CASSANDRA-18344

jacek-lewandowski committed:
Fix statement validation against partition range queries

patch by Jacek Lewandowski; reviewed by Caleb Rackliffe for CASSANDRA-18240

dcapwell committed:
CEP-15: (C*) Add git submodule for Accord

patch by David Capwell; reviewed by Caleb Rackliffe, Michael Semb Wever for CASSANDRA-18204

dcapwell committed:
Switched away from https endpoints for git submodules in favor of relative paths during development

patch by David Capwell; reviewed by Caleb Rackliffe, Michael Semb Wever for CASSANDRA-18204

dcapwell committed:
Add support for prepared statements for accord transactions

patch by David Capwell; reviewed by Alex Petrov, Caleb Rackliffe, Jacek Lewandowski for CASSANDRA-18299

dcapwell committed:
Remove git hook for pre-push as it is redundant and causes issues when merging to mainline

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18309

dcapwell committed:
CEP-15: (Accord) Migrate Accord away from JDK random to a new interface RandomSource

patch by David Capwell; reviewed by Blake Eggleston for CASSANDRA-18213

jacek-lewandowski committed:
Improve transaction statement validation

patch by Jacek Lewandowski; reviewed by David Capwell and Caleb Rackliffe for CASSANDRA-18302

dcapwell committed:
Operations.migrateReadRequiredOperations fails due to concurrent access when TransactionStatement is prepared

patch by David Capwell; reviewed by Ariel Weisberg, Caleb Rackliffe for CASSANDRA-18337

maedhroz committed:
post-rebase fix to adjust for CASSANDRA-17056 in SchemaKeyspace.byteBufferToRange()

maedhroz committed:
Change Accord to 649b0f6: add shareable APPLIED and INVALIDATED implementations of Result

patch by Caleb Rackliffe; reviewed by David Capwell and ? for CASSANDRA-18355

maedhroz committed:
Avoid serializing full TxnData instances to Accord state tables

patch by Caleb Rackliffe; reviewed by David Capwell and ? for CASSANDRA-18355

maedhroz commented:
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-18355

maedhroz commented on .gitmodules:
```diff
@@ -1,4 +1,4 @@
-	url = https://github.com/apache/cassandra-accord.git
-	branch = trunk
+	url = ../cassandra-accord.git
+	branch = CASSANDRA-18355
```
This obv. reverts before committing.

maedhroz commented on src/java/org/apache/cassandra/service/accord/AccordKeyspace.java:
```diff
@@ -652,8 +646,9 @@ public static Command loadCommand(AccordCommandStore commandStore, TxnId txnId)
                     return Command.SerializerSupport.committed(attributes, status, executeAt, promised, accepted, waitingOnCommit, waitingOnApply);
                 case PreApplied:
                 case Applied:
+                    return Command.SerializerSupport.executed(attributes, status, executeAt, promised, accepted, waitingOnCommit, waitingOnApply, writes, Result.APPLIED);
```
Nobody does anything w/ this right now anyway, but I don't know if we need to distinguish between pre-applied and applied...

dcapwell replied:
`org.apache.cassandra.service.accord.AccordObjectSizes#command` needs to be updated right?

dcapwell replied:
think serializing `accord.messages.CheckStatus.CheckStatusOkFull#result` also needs to be updated?  Which should be `org.apache.cassandra.io.IVersionedSerializer#serialize`

maedhroz commented on src/java/org/apache/cassandra/service/accord/serializers/ApplySerializers.java:
```diff
@@ -52,7 +50,7 @@ public Apply deserializeBody(DataInputPlus in, int version, TxnId txnId, Partial
                                                      DepsSerializer.partialDeps.deserialize(in, version),
                                                      CommandSerializers.writes.deserialize(in, version),
-                                                     TxnData.serializer.deserialize(in, version));
+                                                     null);
```
Wondering if we should just remove the `Result` argument from `create()` entirely...

dcapwell replied:
so `Apply` only touches this by passing it to `accord.local.SafeCommand#preapplied`, which calls `accord.local.Command.Executed#executed(accord.local.CommonAttributes, accord.local.SaveStatus, accord.primitives.Timestamp, accord.primitives.Ballot, accord.primitives.Ballot, accord.local.Command.WaitingOn, accord.primitives.Writes, accord.api.Result)`... Given the fact that this case maps to `Result.APPLIED` later on I think its fine to drop, internally we can still pass `Result.APPLIED` in

@belliottsmith thoughts?

maedhroz replied:
After making that change, I'm getting another simulator failure: https://app.circleci.com/pipelines/github/maedhroz/cassandra/709/workflows/221592d1-1ba2-4df3-a22a-c51ae76edd08/jobs/7699/tests#failed-test-0

Will see if I can reproduce locally here...

dcapwell replied:
```
Caused by: java.lang.IllegalArgumentException: Unexpected state FINISHED
	at accord.utils.Invariants.illegalArgument(Invariants.java:54)
	at accord.utils.Invariants.checkArgument(Invariants.java:202)
	at org.apache.cassandra.service.accord.async.AsyncOperation.fail(AsyncOperation.java:182)
	at org.apache.cassandra.service.accord.async.AsyncOperation.run(AsyncOperation.java:273)
	at org.apache.cassandra.service.accord.async.AsyncOperation.callback(AsyncOperation.java:157)
	at accord.utils.async.AsyncCallbacks.lambda$inExecutor$0(AsyncCallbacks.java:31)
```

that's a known issue

dcapwell replied:
[CASSANDRA-18377](https://issues.apache.org/jira/browse/CASSANDRA-18377): CEP-15 (Accord) AsyncOperation can not fail as it has already reached FINISHED

dcapwell committed:
CEP-15 (Accord) Expected reply message with verb ACCORD_INFORM_OF_TXNID_RSP but got ACCORD_SIMPLE_RSP

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18375

dcapwell committed:
CEP-15 (Accord) Original and recover coordinators may hit a race condition with PreApply where reads and writes are interleaved, causing one of the coordinators to see the writes from the other

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18422

belliottsmith committed:
(Accord only) Permit nodes to join a cluster without the full transaction history

patch by Benedict; reviewed by Blake Eggleston for CASSANDRA-18523

dcapwell committed:
CEP-15: (C*) Accord message processing should avoid being passed on to a Stage and run directly in the messageing handler

patch by David Capwell; reviewed by Ariel Weisberg, Benedict Elliott Smith for CASSANDRA-18364

dcapwell committed:
CEP-15: (C*) Enhance in-memory FileSystem to work with mmap and support tests to add custom logic

patch by David Capwell; reviewed by Caleb Rackliffe for CASSANDRA-18485

dcapwell committed:
Ninja for CASSANDRA-18523: Switch accord back to trunk and not pointing to a non-existing bootstrap branch

dcapwell committed:
Ninja for CASSANDRA-18523: bump accord to pick up generator fix for txn_id

dcapwell committed:
CEP-15: (C*) Add notion of CommandsForRanges and make this durable in C*

patch by David Capwell; reviewed by Ariel Weisberg for CASSANDRA-18519

iamaleksey committed:
CEP-15: Extend Accord MessageType with a side effect flag

patch by Aleksey Yeschenko; reviewed by Benedic Elliott Smith for
CASSANDRA-18561

iamaleksey committed:
CEP-15: Convert AccordStateCache cache from write-through to write-back

patch by Aleksey Yeschenko; reviewed by Blake Eggleston for
CASSANDRA-18563

maedhroz commented:
Note: Locally, `AsyncLoaderTest` and `AsyncOperationTest` seem to pass for me.

maedhroz committed:
Change Accord to ce89da2: add shareable APPLIED and INVALIDATED implementations of Result

patch by Caleb Rackliffe; reviewed by ? for CASSANDRA-18355

maedhroz committed:
PAID CircleCI

maedhroz commented:
Committed as https://github.com/apache/cassandra/commit/56e218638db6816f9a6fa66fa2b6fc35977a92a2
